Feat/assert stripped

swrenn opened the pull request:
# Contributor checklist

Leave anything that you believe does not apply unchecked.

- [ ] Bug fixes include regression tests
- [ ] Chores
- [ ] Documentation changes
- [x] Features include unit/acceptance tests
- [ ] Refactoring
- [ ] Update dependencies

zachdaniel commented on lib/ash/test.ex:
```diff
@@ -208,4 +208,82 @@ defmodule Ash.Test do
+
+  defmacro assert_stripped(expr) do
+    quote do
+      raise Ash.Error.Framework.Unsupported,
```
I think in this case just an `ArgumentError` is fine. This error would never be raised from application code/need to be handled in any special way.

zachdaniel commented:
> Also, perhaps I should import ExUnit.Assertions.assert/1 in lib/ash/test.ex, as the code delegates to assert/1?

Yes, good idea 👍

zachdaniel commented:
Your import change is good, but we actually have to do it inside the `quote` blocks for it to be effective for those cases 😄

swrenn commented:
> Your import change is good, but we actually have to do it inside the `quote` blocks for it to be effective for those cases 😄

Whoops 🤦‍♂️. I added fully-qualified calls to `ExUnit.Assertions.assert/1` and reverted the import change so this PR isn't touching unrelated code.

zachdaniel commented on lib/ash/test.ex:
```diff
@@ -208,4 +208,84 @@ defmodule Ash.Test do
+
+  defmacro assert_stripped({:in, _meta, [left, right]}) do
+    quote do
+      ExUnit.Assertions.assert(strip_metadata(unquote(left)) in strip_metadata(unquote(right)))
```
Even w/ fully qualified calls, its a macro, so these macros should either import or require `ExUnit.Assertions`

swrenn replied:
I apologize for my noob mistakes. What is the impact of injecting import statements at every call site? Is that the right solution? Is there a case here for `__using__`?

zachdaniel commented on .formatter.exs:
```diff
@@ -25,6 +25,7 @@ spark_locals_without_parens = [
   ash_step: 1,
   ash_step: 2,
   ash_step: 3,
+  assert_stripped: 1,
```
Ah, just noticed this. This will need to go in a separate list below, and get added to this list. This list of locals is generated from the DSL definitions.

Then its good 😄

swrenn replied:
Do you see something like this often in Elixir code?

```
locals_without_parens = [
  {:assert_stripped, 1}
  | spark_locals_without_parens
]
```

Or is this better?

```
other_locals_without_parens = [
  assert_stripped: 1
]

locals_without_parens =
  spark_locals_without_parens ++
    other_locals_without_parens
```

zachdaniel commented:
Try it and see 😄 `mix docs && open doc/index.html`

zachdaniel commented:
🚀 Thank you for your contribution! 🚀
